ballerina-to-openapi/src/main/java/io/ballerina/openapi/service/mapper/hateoas/HateoasMetadataVisitor.java

```diff
@@ -0,0 +1,91 @@
+import io.ballerina.compiler.api.symbols.ServiceDeclarationSymbol;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.openapi.service.mapper.utils.MapperCommonUtils;
+
+import java.util.Optional;
+
+import static io.ballerina.openapi.service.mapper.hateoas.HateoasContextHolder.getHateoasContextHolder;
+import static io.ballerina.openapi.service.mapper.utils.MapperCommonUtils.getResourceConfigAnnotation;
+import static io.ballerina.openapi.service.mapper.utils.MapperCommonUtils.getValueForAnnotationFields;
+
+/**
+ * Visitor to retrieve the Hateoas meta-data from resource functions within a {@link ServiceDeclarationNode}.
+ *
+ * @since 1.6.0
+ */
+public class HateoasMetadataVisitor extends NodeVisitor {
+    private final String packageId;
+    private final SemanticModel semanticModel;
+
+    public HateoasMetadataVisitor(String packageId, SemanticModel semanticModel) {
+        this.packageId = packageId;
+        this.semanticModel = semanticModel;
+    }
+
+    @Override
+    public void visit(ServiceDeclarationNode serviceNode) {
+        boolean isHttpService = MapperCommonUtils.isHttpService(serviceNode, semanticModel);
+        if (!isHttpService) {
+            return;
+        }
+        Optional<Symbol> serviceDeclarationOpt = semanticModel.symbol(serviceNode);
+        if (serviceDeclarationOpt.isEmpty()) {
+            return;
+        }
+        ServiceDeclarationSymbol serviceSymbol = (ServiceDeclarationSymbol) serviceDeclarationOpt.get();
+        int serviceId = serviceSymbol.hashCode();
+        for (Node child : serviceNode.children()) {
+            if (SyntaxKind.RESOURCE_ACCESSOR_DEFINITION.equals(child.kind())) {
+                FunctionDefinitionNode resourceFunction = (FunctionDefinitionNode) child;
+                String resourceMethod = resourceFunction.functionName().text();
+                String operationId = generateOperationId(resourceFunction);
+                Optional<String> resourceName = getResourceConfigAnnotation(resourceFunction)
+                        .flatMap(resourceConfig -> getValueForAnnotationFields(resourceConfig, "name"));
+                if (resourceName.isEmpty()) {
+                    return;
+                }
+                String cleanedResourceName = resourceName.get().replaceAll("\"", "");
+                Resource hateoasResource = new Resource(resourceMethod, operationId);
+                getHateoasContextHolder().updateHateoasResource(
+                        packageId, serviceId, cleanedResourceName, hateoasResource);
+            }
+        }
+    }
+
+    private String generateOperationId(FunctionDefinitionNode resourceFunction) {
```

Comment:
Yes, the `operationId` is the same with what we set from the `ResourceMapper`

Reply:
Then can we define a method in the ResourceMapper interface and get the operation id from there?
This way if we change the logic for generating operation id in the future, we do not need to change in both components. Also this will imply that the HateoasMapper should know about the ResourceMapper to retrieve the operation id. IMO this makes sense

Reply:
With this change [dfa05455b7fb2180d9b9ece0f445314a25553116](https://github.com/ballerina-platform/openapi-tools/pull/1589/commits/dfa05455b7fb2180d9b9ece0f445314a25553116), 
I refactored the method `getOperationId` in `MapperCommonUtils`  where it can be accessed by both `ResourceMapperImpl` and the `HateoasMetadataVisitor` and get the operationId. I think this should solve our problem. With this refactoring there won't be any duplication of the code. 

Can you review this change.. @ayeshLK @TharmiganK